cpp/src/arrow/filesystem/s3_internal.h

```diff
@@ -88,6 +88,84 @@ inline bool IsAlreadyExists(const Aws::Client::AWSError<Aws::S3::S3Errors>& erro
           error_type == Aws::S3::S3Errors::BUCKET_ALREADY_OWNED_BY_YOU);
 }
 
+std::string S3ErrorToString(int error_type) {
+  switch (static_cast<Aws::S3::S3Errors>(error_type)) {
+    case Aws::S3::S3Errors::INCOMPLETE_SIGNATURE:
+      return "INCOMPLETE_SIGNATURE";
```

Comment:
Hmm, perhaps we can make the boilerplate a bit lighter using a macro? e.g.
```suggestion
#define S3_ERROR_CASE(NAME) \
    case Aws::S3::S3Errors:: ## NAME: \
      return # NAME

    S3_ERROR_CASE(INCOMPLETE_SIGNATURE)
    S3_ERROR_CASE(INTERNAL_FAILURE)
    ...

#undef S3_ERROR_CASE
```